napari/plugins/__init__.py

```diff
@@ -0,0 +1,12 @@
+import pluggy
+from .manager import NapariPluginManager
+
+hookimpl = pluggy.HookimplMarker("napari")
+"""Marker to be imported and used in plugins (and for own implementations)
+Note: plugins may also just import pluggy directly and make their own
+napari hookimpl.
+"""
+
+# a singleton... but doesn't need to be.
```

Comment:
Not quite sure what this comment intends. Is this something that needs to be decided before merge?

Reply:
> I'm not quite sure about leaving the hookspec and our built-in hookimpl for `napari_get_reader` without hooking them up, as I'd like to think about more about how they actually work in practice before committing to them.

here's how it looks to add a path to the viewer (this worked in practice before I removed it from the PR.  the remaining details have more to do with how we want to accept single vs multiple paths):

```python
    def _add_path(self, path):
        if path:
            reader = plugin_manager.hook.napari_get_reader(path=path)
            layer_data = reader(path)
            for data in layer_data:
                self.viewer._add_layer_data(*data)
            self._last_visited_dir = os.path.dirname(path)
```

> right now it seems like you've moved away from the checker / reader approach into a single function that then returns another function if it is readable right? This maybe the way we want to go, but might be nice to be able to review that independently from all the other changes. 

we can certainly revisit that, and I don't think changing the hookspec in a later PR is a problem at all.  also, fwiw, @jni was a fan of that strategy: https://github.com/napari/napari/pull/908#issuecomment-578566907

> If you remove it is there still enough left in this PR (in terms of tests you've added etc) for it to be meaningful? That follow-up PR could add hookspec and our built-in hookimpl for `napari_get_reader` and hook them up in one go

the hookspec is largely there for testings, but I could delay those tests for later... (but they work in the abstract as well, and are a large part of the "meat" of this PR).
but not too attached either way.  give it a think and let me know

Reply:
Ok, I think it's fine to try and get this in now with the minimal reader stuff and press on - we can continue to refine some of these things as we do.

Curious this comment in the code means, it is still not clear to me:

> # a singleton... but doesn't need to be.

Reply:
yeah that can be removed... that was partially an internal comment in response to https://github.com/napari/napari/pull/891#issuecomment-577850347 and related to your https://github.com/napari/napari/pull/908#issuecomment-580804463.  I'll remove it.